comms/dht/src/outbound/encryption.rs

```diff
@@ -79,50 +91,66 @@ where S: Service<DhtOutboundMessage, Response = (), Error = PipelineError> + Clo
         Poll::Ready(Ok(()))
     }
 
-    fn call(&mut self, msg: DhtOutboundMessage) -> Self::Future {
-        Self::handle_message(self.inner.clone(), Arc::clone(&self.node_identity), msg)
+    fn call(&mut self, mut message: DhtOutboundMessage) -> Self::Future {
+        let next_service = self.inner.clone();
+        let node_identity = Arc::clone(&self.node_identity);
+        async move {
+            trace!(target: LOG_TARGET, "DHT Message flags: {:?}", message.dht_flags);
+            match &message.encryption {
+                OutboundEncryption::EncryptFor(public_key) => {
+                    debug!(target: LOG_TARGET, "Encrypting message for {}", public_key);
+                    // Generate ephemeral public/private key pair and ECDH shared secret
+                    let (e_sk, e_pk) = CommsPublicKey::random_keypair(&mut OsRng);
+                    let shared_ephemeral_secret = crypt::generate_ecdh_secret(&e_sk, &**public_key);
+                    // Encrypt the message with the body
+                    let encrypted_body =
+                        crypt::encrypt(&shared_ephemeral_secret, &message.body).map_err(PipelineError::from_debug)?;
+
+                    // Sign the encrypted message
+                    let origin_mac = create_origin_mac(&node_identity, &encrypted_body)?;
```

Comment:
I'm not sure there is a solution, but the detection and banning of a spamming node will be difficult with the encryption of the source identity using the shared ephemeral secret.

Reply:
The general idea is that peers will reject messages over a certain size and temporarily reject/rate limit these messages if too many are received limiting the impact of spam messages. However, the heuristics for this are not completely clear. It's worth noting that even with the source public key in the header a node could still spam by using an ephemeral public key when sending the message.

Reply:
We may also want to establish bonded relationships with peers before allowing them to send/propagate messages of this type.